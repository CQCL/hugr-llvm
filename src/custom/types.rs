use itertools::Itertools as _;

use hugr::types::CustomType;

use anyhow::{bail, Result};
use inkwell::types::{BasicMetadataTypeEnum, BasicType as _, BasicTypeEnum, FunctionType};

use crate::{
    sum::LLVMSumType,
    types::{HugrFuncType, HugrSumType, HugrType, TypingSession},
    utils::type_map::TypeMapping,
};

pub trait LLVMCustomTypeFn<'a>:
    for<'c> Fn(TypingSession<'c>, &CustomType) -> Result<BasicTypeEnum<'c>> + 'a
{
}

impl<
        'a,
        F: for<'c> Fn(TypingSession<'c>, &CustomType) -> Result<BasicTypeEnum<'c>> + 'a + ?Sized,
    > LLVMCustomTypeFn<'a> for F
{
}

#[derive(Default, Clone)]
pub struct LLVMTypeMapping;

impl TypeMapping for LLVMTypeMapping {
    type InV<'c> = TypingSession<'c>;

    type OutV<'c> = BasicTypeEnum<'c>;

    type SumOutV<'c> = LLVMSumType<'c>;

    type FuncOutV<'c> = FunctionType<'c>;

    fn sum_into_out<'c>(&self, sum: Self::SumOutV<'c>) -> Self::OutV<'c> {
        sum.as_basic_type_enum()
    }

    fn func_into_out<'c>(&self, sum: Self::FuncOutV<'c>) -> Self::OutV<'c> {
        sum.ptr_type(Default::default()).as_basic_type_enum()
    }

    fn default_out<'c>(&self, hugr_type: &HugrType) -> Result<Self::OutV<'c>> {
        bail!("Unsupported type: {hugr_type}")
    }

    fn map_sum_type<'c>(
        &self,
        sum_type: &HugrSumType,
        context: TypingSession<'c>,
        variants: impl IntoIterator<Item = Vec<Self::OutV<'c>>>,
    ) -> Result<Self::SumOutV<'c>> {
        LLVMSumType::try_new2(
            context.iw_context(),
            variants.into_iter().collect(),
            sum_type.clone(),
        )
    }

    fn map_function_type<'c>(
        &self,
        _: &HugrFuncType,
        context: TypingSession<'c>,
        inputs: impl IntoIterator<Item = Self::OutV<'c>>,
        outputs: impl IntoIterator<Item = Self::OutV<'c>>,
    ) -> Result<Self::FuncOutV<'c>> {
        let iw_context = context.iw_context();
        let inputs: Vec<BasicMetadataTypeEnum<'c>> = inputs.into_iter().map_into().collect_vec();
        let outputs = outputs.into_iter().collect_vec();
        Ok(match outputs.as_slice() {
            &[] => iw_context.void_type().fn_type(&inputs, false),
            [res] => res.fn_type(&inputs, false),
            ress => iw_context.struct_type(ress, false).fn_type(&inputs, false),
        })
    }
}
