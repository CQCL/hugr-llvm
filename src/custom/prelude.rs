use hugr::{extension::prelude, HugrView};
use inkwell::types::BasicType;

use super::{CodegenExtension, CodegenExtsMap};

struct PreludeCodegenExtension;

impl<'c, H: HugrView> CodegenExtension<'c, H> for PreludeCodegenExtension {
    fn extension(&self) -> hugr::extension::ExtensionId {
        return prelude::PRELUDE_ID
    }

    fn llvm_type(
        &self,
        context: &crate::types::TypingSession<'c, H>,
        hugr_type: &hugr::types::CustomType,
    ) -> anyhow::Result<inkwell::types::BasicTypeEnum<'c>> {
        match hugr_type.name().as_str() {
            "qubit" => Ok(context.iw_context().i32_type().as_basic_type_enum()),
            _ => todo!(),
        }
    }

    fn emitter<'a>(
        &self,
        context: &'a mut crate::emit::func::EmitFuncContext<'c, H>,
    ) -> Box<dyn crate::emit::EmitOp<'c, hugr::ops::CustomOp, H> + 'a> {
        todo!()
    }
}

pub fn add_prelude_extensions<H: HugrView>(cem: CodegenExtsMap<'_, H>) -> CodegenExtsMap<'_, H> {
    cem.add_cge(PreludeCodegenExtension)
}

impl<H: HugrView> CodegenExtsMap<'_, H> {
    pub fn add_prelude_extensions(self) -> Self {
        add_prelude_extensions(self)
    }
}
