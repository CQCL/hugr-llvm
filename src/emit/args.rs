use hugr::{ops::OpType, HugrView};
use inkwell::values::BasicValueEnum;

use crate::fat::FatNode;

use super::func::RowPromise;

/// A type used whenever emission is delegated to a function
pub struct EmitOpArgs<'c, OT, H> {
    /// This [HugrView] and [Node] we are emitting
    pub node: FatNode<'c, OT, H>,
    /// The values that should be used for all Value input ports of the node
    pub inputs: Vec<BasicValueEnum<'c>>,
    /// The results of the node should be put here
    pub outputs: RowPromise<'c>,
}

impl<'c, OT, H> EmitOpArgs<'c, OT, H> {
    /// Get the internal [FatNode]
    pub fn node(&self) -> FatNode<'c, OT, H> {
        self.node.clone()
    }
}

impl<'c, H: HugrView> EmitOpArgs<'c, OpType, H> {
    /// Attempt to specialise the internal [FatNode].
    pub fn try_into_ot<OT: 'c>(self) -> Result<EmitOpArgs<'c, OT, H>, Self>
    where
        for<'a> &'a OpType: TryInto<&'a OT>,
    {
        let EmitOpArgs {
            node,
            inputs,
            outputs,
        } = self;
        match node.try_into_ot() {
            Some(new_node) => Ok(EmitOpArgs {
                node: new_node,
                inputs,
                outputs,
            }),
            None => Err(EmitOpArgs {
                node,
                inputs,
                outputs,
            }),
        }
    }

    /// Specialise the internal [FatNode].
    ///
    /// Panics if `OT` is not the `get_optype` of the internal [Node].
    pub fn into_ot<OTInto: PartialEq + 'c>(self, ot: &OTInto) -> EmitOpArgs<'c, OTInto, H>
    where
        for<'a> &'a OpType: TryInto<&'a OTInto>,
    {
        let EmitOpArgs {
            node,
            inputs,
            outputs,
        } = self;
        EmitOpArgs {
            node: node.as_ot(ot),
            inputs,
            outputs,
        }
    }
}

impl<'c, OT: 'c, H: HugrView> AsRef<OT> for EmitOpArgs<'c, OT, H>
where
    for<'a> &'a OpType: TryInto<&'a OT>,
{
    fn as_ref(&self) -> &OT {
        self.node.as_ref()
    }
}
