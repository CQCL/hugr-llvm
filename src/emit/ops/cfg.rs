use std::collections::HashMap;

use anyhow::{anyhow, Result};
use hugr::{
    ops::{DataflowBlock, ExitBlock, OpType, CFG},
    types::SumType,
    HugrView, NodeIndex,
};
use inkwell::{basic_block::BasicBlock, values::BasicValueEnum};
use itertools::Itertools as _;

use crate::{
    emit::{
        func::{EmitFuncContext, RowMailBox, RowPromise},
        EmitOp, EmitOpArgs,
    },
    fat::FatNode,
};

use super::emit_dataflow_parent;

pub struct CfgEmitter<'c, 'd, H: HugrView> {
    context: &'d mut EmitFuncContext<'c, H>,
    bbs: HashMap<FatNode<'c, OpType, H>, (BasicBlock<'c>, RowMailBox<'c>)>,
    inputs: Option<Vec<BasicValueEnum<'c>>>,
    outputs: Option<RowPromise<'c>>,
    node: FatNode<'c, CFG, H>,
    entry_node: FatNode<'c, DataflowBlock, H>,
    exit_node: FatNode<'c, ExitBlock, H>,
}

impl<'c, 'd, H: HugrView> CfgEmitter<'c, 'd, H> {
    // Constructs a new CfgEmitter. Creates a basic block for each of
    // the children in the llvm function. Note that this does not move the
    // position of the builder.
    pub fn new(
        context: &'d mut EmitFuncContext<'c, H>,
        args: EmitOpArgs<'c, CFG, H>,
    ) -> Result<Self> {
        let node = args.node();
        let (inputs, outputs) = (Some(args.inputs), Some(args.outputs));

        // create this now so that it will be the last block and we can use it
        // to crate the other blocks immediately before it. This is just for
        // nice block ordering.
        let exit_block = context.new_basic_block("", None);
        let bbs = node
            .children()
            .map(|child| {
                if child.is_exit_block() {
                    let output_row = {
                        let out_types = node.out_value_types().map(|x| x.1).collect_vec();
                        context.new_row_mail_box(out_types.iter(), "")?
                    };
                    Ok((child, (exit_block, output_row)))
                } else {
                    let bb = context.new_basic_block("", Some(exit_block));
                    let (i, _) = child.get_io().unwrap();
                    Ok((child, (bb, context.node_outs_rmb(i)?)))
                }
            })
            .collect::<Result<HashMap<_, _>>>()?;
        let (entry_node, exit_node) = node.get_entry_exit().unwrap();
        Ok(CfgEmitter {
            context,
            bbs,
            node,
            inputs,
            outputs,
            entry_node,
            exit_node,
        })
    }

    fn take_inputs(&mut self) -> Result<Vec<BasicValueEnum<'c>>> {
        self.inputs.take().ok_or(anyhow!("Couldn't take inputs"))
    }

    fn take_outputs(&mut self) -> Result<RowPromise<'c>> {
        self.outputs.take().ok_or(anyhow!("Couldn't take inputs"))
    }

    fn get_block_data<OT>(
        &self,
        node: &FatNode<'c, OT, H>,
    ) -> Result<&(BasicBlock<'c>, RowMailBox<'c>)>
    where
        OT: Into<OpType> + 'c,
    {
        self.bbs
            .get(&node.clone().generalise())
            .ok_or(anyhow!("Couldn't get block data for: {}", node.index()))
    }

    /// Consume the emitter by emitting each child of the node.
    /// After returning the builder will be at the end of the exit block.
    pub fn emit_children(mut self) -> Result<()> {
        // write the inputs of the cfg node into the inputs of the entry
        // dataflowblock node, and then branch to the basic block of that entry
        // node.
        let inputs = self.take_inputs()?;
        let (entry_bb, inputs_rmb) = self.get_block_data(&self.entry_node).cloned()?;
        let builder = self.context.builder();
        inputs_rmb.write(builder, inputs)?;
        builder.build_unconditional_branch(entry_bb)?;

        // emit each child by delegating to the `impl EmitOp<_>` of self.
        for c in self.node.children() {
            let (inputs, outputs) = (vec![], RowMailBox::new_empty().promise());
            if let Some(node) = c.try_into_ot::<DataflowBlock>() {
                self.emit(EmitOpArgs {
                    node,
                    inputs,
                    outputs,
                })?;
            } else if let Some(node) = c.try_into_ot::<ExitBlock>() {
                self.emit(EmitOpArgs {
                    node,
                    inputs,
                    outputs,
                })?;
            } else {
                Err(anyhow!("unknown optype: {c}"))?;
            }
        }

        // move the builder to the end of the exit block
        let (exit_bb, _) = self.get_block_data(&self.exit_node).cloned()?;
        self.context.builder().position_at_end(exit_bb);
        Ok(())
    }
}

impl<'c, H: HugrView> EmitOp<'c, DataflowBlock, H> for CfgEmitter<'c, '_, H> {
    fn emit(
        &mut self,
        EmitOpArgs {
            node,
            inputs: _,
            outputs: _,
        }: EmitOpArgs<'c, DataflowBlock, H>,
    ) -> Result<()> {
        // our entry basic block and our input RowMailBox
        let (bb, inputs_rmb) = self.bbs.get(&node.clone().generalise()).unwrap();
        // the basic block and mailbox of each of our successors
        let successor_data = node
            .output_neighbours()
            .map(|succ| self.get_block_data(&succ).map(|x| x.clone()))
            .collect::<Result<Vec<_>>>()?;

        self.context.build_positioned(*bb, |context| {
            let (_, o) = node.get_io().unwrap();
            // get the rowmailbox for our output node
            let outputs_rmb = context.node_ins_rmb(o)?;
            // read the values from our input node
            let inputs = inputs_rmb.read_vec(context.builder(), [])?;

            // emit all our children and read the values from the rowmailbox of our output node
            emit_dataflow_parent(
                context,
                EmitOpArgs {
                    node: node.clone(),
                    inputs,
                    outputs: outputs_rmb.promise(),
                },
            )?;
            let outputs = outputs_rmb.read_vec(context.builder(), [])?;

            let branch_sum_type = SumType::new(node.sum_rows.clone());
            let llvm_sum_type = context.llvm_sum_type(branch_sum_type)?;
            let tag_bbs = successor_data
                .into_iter()
                .enumerate()
                .map(|(tag, (target_bb, target_rmb))| {
                    let bb = context.build_positioned_new_block("", Some(*bb), |context, bb| {
                        let builder = context.builder();
                        let mut vals =
                            llvm_sum_type.build_untag(builder, tag as u32, outputs[0])?;
                        vals.extend(&outputs[1..]);
                        target_rmb.write(builder, vals)?;
                        builder.build_unconditional_branch(target_bb)?;
                        Ok::<_, anyhow::Error>(bb)
                    })?;
                    Ok((
                        llvm_sum_type.get_tag_type().const_int(tag as u64, false),
                        bb,
                    ))
                })
                .collect::<Result<Vec<_>>>()?;
            let tag_v = llvm_sum_type.build_get_tag(context.builder(), outputs[0])?;
            context
                .builder()
                .build_switch(tag_v, tag_bbs[0].1, &tag_bbs[1..])?;
            Ok(())
        })
    }
}

impl<'c, H: HugrView> EmitOp<'c, ExitBlock, H> for CfgEmitter<'c, '_, H> {
    fn emit(&mut self, args: EmitOpArgs<'c, ExitBlock, H>) -> Result<()> {
        let outputs = self.take_outputs()?;
        let (bb, inputs_rmb) = self.bbs.get(&args.node().generalise()).unwrap();
        self.context.build_positioned(*bb, |context| {
            let builder = context.builder();
            outputs.finish(builder, inputs_rmb.read_vec(builder, [])?)
        })
    }
}
