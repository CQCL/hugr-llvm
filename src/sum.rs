use crate::types::{HugrSumType, TypingSession};

use anyhow::{anyhow, Result};
use hugr::types::TypeRow;
use inkwell::{
    builder::Builder,
    types::{AnyType, AsTypeRef, BasicType, BasicTypeEnum, IntType, StructType},
    values::{BasicValue, BasicValueEnum, IntValue, StructValue},
};
use itertools::{zip_eq, Itertools};
use llvm_sys_140::prelude::LLVMTypeRef;

/// The opaque representation of a hugr [SumType].
///
/// Using the public methods of this type one emit "tag"s,"untag"s, and
/// "get_tag"s while not exposing the underlying LLVM representation.
///
/// We offer impls of [BasicType] and parent traits.
#[derive(Debug)]
pub struct LLVMSumType<'c>(StructType<'c>, HugrSumType);

impl<'c> LLVMSumType<'c> {
    /// Attempt to create a new `LLVMSumType` from a [HugrSumType].
    pub fn try_new<H>(session: &TypingSession<'c, H>, sum_type: HugrSumType) -> Result<Self> {
        assert!(sum_type.num_variants() < u32::MAX as usize);
        let variants = (0..sum_type.num_variants())
            .map(|i| {
                let tr = Self::get_variant_typerow_st(&sum_type, i as u32)?;
                tr.iter()
                    .map(|t| session.llvm_type(t))
                    .collect::<Result<Vec<_>>>()
            })
            .collect::<Result<Vec<_>>>()?;
        let has_tag_field = Self::sum_type_has_tag_field(&sum_type);
        let types = has_tag_field
            .then_some(session.iw_context().i32_type().as_basic_type_enum())
            .into_iter()
            .chain(
                variants
                    .iter()
                    .map(|lty_vec| session.iw_context().struct_type(lty_vec, false).into()),
            )
            .collect_vec();
        Ok(Self(
            session.iw_context().struct_type(&types, false),
            sum_type.clone(),
        ))
    }

    /// Returns an LLVM constant value of `undef`.
    pub fn get_undef(&self) -> impl BasicValue<'c> {
        self.0.get_undef()
    }

    /// Returns an LLVM constant value of `poison`.
    pub fn get_poison(&self) -> impl BasicValue<'c> {
        self.0.get_poison()
    }

    /// Emit instructions to read the tag of a value of type `LLVMSumType`.
    ///
    /// The type of the value is that returned by [LLVMSumType::get_tag_type].
    pub fn build_get_tag(
        &self,
        builder: &Builder<'c>,
        v: impl BasicValue<'c>,
    ) -> Result<IntValue<'c>> {
        let struct_value: StructValue<'c> = v
            .as_basic_value_enum()
            .try_into()
            .map_err(|_| anyhow!("Not a struct type"))?;
        if self.has_tag_field() {
            Ok(builder
                .build_extract_value(struct_value, 0, "")?
                .into_int_value())
        } else {
            Ok(self.get_tag_type().const_int(0, false))
        }
    }

    /// Emit instructions to read the inner values of a value of type
    /// `LLVMSumType`, on the assumption that it's tag is `tag`.
    ///
    /// If it's tag is not `tag`, the returned values will be poison.
    pub fn build_untag(
        &self,
        builder: &Builder<'c>,
        tag: u32,
        v: impl BasicValue<'c>,
    ) -> Result<Vec<BasicValueEnum<'c>>> {
        debug_assert!((tag as usize) < self.1.num_variants());
        debug_assert!(v.as_basic_value_enum().get_type() == self.0.as_basic_type_enum());

        let v: StructValue<'c> = builder
            .build_extract_value(
                v.as_basic_value_enum().into_struct_value(),
                self.get_variant_index(tag),
                "",
            )?
            .into_struct_value();
        let r = (0..v.get_type().count_fields())
            .map(|i| Ok(builder.build_extract_value(v, i, "")?.as_basic_value_enum()))
            .collect::<Result<Vec<_>>>()?;
        debug_assert_eq!(r.len(), self.num_fields(tag).unwrap());
        Ok(r)
    }

    /// Emit instructions to build a value of type `LLVMSumType`, being of variant `tag`.
    pub fn build_tag(
        &self,
        builder: &Builder<'c>,
        tag: u32,
        vs: Vec<BasicValueEnum<'c>>,
    ) -> Result<BasicValueEnum<'c>> {
        let expected_num_fields = self.num_fields(tag)?;
        if expected_num_fields != vs.len() {
            Err(anyhow!("LLVMSumType::build: wrong number of fields: expected: {expected_num_fields} actual: {}", vs.len()))?
        }
        let variant_index = self.get_variant_index(tag);
        let row_t = self
            .0
            .get_field_type_at_index(variant_index)
            .ok_or(anyhow!("LLVMSumType::build: no field type at index"))
            .and_then(|row_t| {
                if !row_t.is_struct_type() {
                    Err(anyhow!("LLVMSumType::build"))?
                }
                Ok(row_t.into_struct_type())
            })?;
        debug_assert!(zip_eq(vs.iter(), row_t.get_field_types().into_iter())
            .all(|(lhs, rhs)| lhs.as_basic_value_enum().get_type() == rhs));
        let mut row_v = row_t.get_undef();
        for (i, val) in vs.into_iter().enumerate() {
            row_v = builder
                .build_insert_value(row_v, val, i as u32, "")?
                .into_struct_value();
        }
        let mut sum_v = self.get_poison().as_basic_value_enum().into_struct_value();
        if self.has_tag_field() {
            sum_v = builder
                .build_insert_value(
                    sum_v,
                    self.get_tag_type().const_int(tag as u64, false),
                    0u32,
                    "",
                )?
                .into_struct_value();
        }
        Ok(builder
            .build_insert_value(sum_v, row_v, variant_index, "")?
            .as_basic_value_enum())
    }

    /// Get the type of the value that would be returned by `build_get_tag`.
    pub fn get_tag_type(&self) -> IntType<'c> {
        self.0.get_context().i32_type()
    }

    fn sum_type_has_tag_field(st: &HugrSumType) -> bool {
        st.num_variants() >= 2
    }

    fn has_tag_field(&self) -> bool {
        Self::sum_type_has_tag_field(&self.1)
    }

    fn get_variant_index(&self, tag: u32) -> u32 {
        tag + (if self.has_tag_field() { 1 } else { 0 })
    }

    fn get_variant_typerow_st(sum_type: &HugrSumType, tag: u32) -> Result<TypeRow> {
        sum_type
            .get_variant(tag as usize)
            .ok_or(anyhow!("Bad variant index {tag} in {sum_type}"))
            .and_then(|tr| Ok(TypeRow::try_from(tr.clone())?))
    }

    fn get_variant_typerow(&self, tag: u32) -> Result<TypeRow> {
        Self::get_variant_typerow_st(&self.1, tag)
    }

    fn num_fields(&self, tag: u32) -> Result<usize> {
        let tr = self.get_variant_typerow(tag)?;
        Ok(tr.len())
    }
}

impl<'c> From<LLVMSumType<'c>> for BasicTypeEnum<'c> {
    fn from(value: LLVMSumType<'c>) -> Self {
        value.0.as_basic_type_enum()
    }
}

impl<'c> std::fmt::Display for LLVMSumType<'c> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.0.fmt(f)
    }
}

unsafe impl<'c> AsTypeRef for LLVMSumType<'c> {
    fn as_type_ref(&self) -> LLVMTypeRef {
        self.0.as_type_ref()
    }
}

unsafe impl<'c> AnyType<'c> for LLVMSumType<'c> {}

unsafe impl<'c> BasicType<'c> for LLVMSumType<'c> {}
