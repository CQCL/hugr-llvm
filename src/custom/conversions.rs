use hugr::extension::ExtensionId;
use hugr::types::{CustomType};
use llvm_sys_140::LLVMValue;

use crate::emit::{
    func::EmitFuncContext, ops::emit_custom_unary_op,
    EmitOp, EmitOpArgs,
};

use hugr::std_extensions::arithmetic::conversions::ConvertOpDef;

use super::{CodegenExtension, CodegenExtsMap};
use crate::types::TypingSession;
use anyhow::{anyhow, Result};

use hugr::{HugrView, std_extensions::arithmetic::{conversions, int_types::INT_TYPES}, ops::CustomOp};

use hugr::extension::simple_op::MakeExtensionOp;
use hugr::types::SumType;
use hugr::extension::prelude::ERROR_TYPE;

use inkwell::types::BasicTypeEnum;
use inkwell::values::{AnyValue, BasicValue, BasicValueEnum};

use crate::emit::ops::emit_value;

use hugr::ops::constant::Value;

use inkwell::builder::Builder;

use crate::sum::LLVMSumType;

struct ConversionsEmitter<'c, 'd, H>(&'d mut EmitFuncContext<'c, H>);

/*
fn emit_trunc_op<'c>(builder: Builder<'c>, int_type: BasicTypeEnum, float_value: BasicValueEnum) -> Result<()> {
    // Make a maximum size int and convert it to float to work out if the
    // truncation should succeed.
    let x = builder.build_float_to_unsigned_int(float_value.into_float_value(), int_type.into_int_type(), "")?;
    // If x is poison, we need to return a failure tag
    if x.is_poison() {
        emit_value(builder., &Value::false_val())?;
    }
    // Otherwise, we need to pack the result in a success tag
    else {
        todo!()
    }

    Ok(())
}
 */


fn truncate_output(width: u8) -> SumType {
    SumType::new([
        INT_TYPES[width as usize].clone(),
        ERROR_TYPE,
    ])
}


impl<'c, H: HugrView> EmitOp<'c, CustomOp, H> for ConversionsEmitter<'c, '_, H> {
    fn emit(&mut self, args: EmitOpArgs<'c, CustomOp, H>) -> Result<()> {
        let conversion_op = ConvertOpDef::from_optype(&args.node().generalise()).ok_or(anyhow!(
            "ConversionsEmitter from_optype failed: {:?}",
            args.node().as_ref()
        ))?;
        match conversion_op {
            ConvertOpDef::trunc_u => {
                // TODO: Be generic over the width - work it out from what's on the hugr node
                let sum_ty = LLVMSumType::try_new(&self.0.typing_session(), truncate_output(5))?;
                emit_custom_unary_op(self.0, args, |builder, arg, out_tys| {
                    // TODO proper error handling
                    let [out_ty] = out_tys else { panic!(""); };
                    let int_ty = out_ty.into_struct_type().get_field_type_at_index(0).unwrap().into_int_type();
                    let trunc_result = builder
                        .build_float_to_unsigned_int(arg.into_float_value(), int_ty, "")?
                        .as_basic_value_enum();
                    let optional_result = if trunc_result.is_poison() {
                        // TODO: Construct an error object
                        sum_ty.build_tag(builder, 1, vec![])
                    } else {
                        sum_ty.build_tag(builder, 0, vec![trunc_result])
                    }?;

                    Ok(vec![optional_result])
                })
            },

            ConvertOpDef::trunc_s => {
                let sum_ty = LLVMSumType::try_new(&self.0.typing_session(), truncate_output(5))?;
                emit_custom_unary_op(self.0, args, |builder, arg, out_tys| {
                    // TODO proper error handling
                    let [out_ty] = out_tys else { panic!(""); };
                    let int_ty = out_ty.into_struct_type().get_field_type_at_index(0).unwrap().into_int_type();
                    let trunc_result = builder
                        .build_float_to_signed_int(arg.into_float_value(), int_ty, "")?
                        .as_basic_value_enum();
                    let optional_result = if trunc_result.is_poison() {
                        // TODO: Construct an error object
                        sum_ty.build_tag(builder, 1, vec![])
                    } else {
                        sum_ty.build_tag(builder, 0, vec![trunc_result])
                    }?;

                    Ok(vec![optional_result])
                })
            },

            ConvertOpDef::convert_u => emit_custom_unary_op(self.0, args, |builder, arg, out_tys| {
                    // TODO proper error handling
                let [out_ty] = out_tys else {  panic!(""); };
                Ok(vec![builder
                    .build_unsigned_int_to_float(arg.into_int_value(), out_ty.into_float_type(), "")?
                    .as_basic_value_enum()])
            }),
            ConvertOpDef::convert_s => emit_custom_unary_op(self.0, args, |builder, arg, out_tys| {
                    // TODO proper error handling
                let [out_ty] = out_tys else { panic!(""); };
                Ok(vec![builder
                    .build_signed_int_to_float(arg.into_int_value(), out_ty.into_float_type(), "")?
                    .as_basic_value_enum()])
            }),
            _ => Err(anyhow!("Conversion op not implemented: {:?}", args.node().as_ref())),
        }
        //Ok(())
    }
}

pub struct ConversionsCodegenExtension;

impl<'c, H: HugrView> CodegenExtension<'c, H> for ConversionsCodegenExtension {
    fn extension(&self) -> ExtensionId {
        conversions::EXTENSION_ID
    }

    fn llvm_type<'d>(
        &self,
        _context: &TypingSession<'c, H>,
        hugr_type: &CustomType,
    ) -> Result<BasicTypeEnum<'c>> {
        Err(anyhow!(
            "IntOpsCodegenExtension: unsupported type: {}",
            hugr_type
        ))
    }

    fn emitter<'a>(
        &self,
        context: &'a mut EmitFuncContext<'c, H>,
    ) -> Box<dyn EmitOp<'c, CustomOp, H> + 'a> {
        Box::new(ConversionsEmitter(context))
    }


}


pub fn add_conversions_extension<H: HugrView>(cem: CodegenExtsMap<'_, H>) -> CodegenExtsMap<'_, H> {
    cem.add_cge(ConversionsCodegenExtension)
}

#[cfg(test)]
mod test {

    use super::*;

    use hugr::extension::prelude::ERROR_TYPE_NAME;
    use rstest::rstest;
    use crate::check_emission;
    use crate::test::{llvm_ctx, TestContext};
    use hugr::{
        builder::{Dataflow, DataflowSubContainer},
        extension::prelude::ERROR_TYPE,
        std_extensions::arithmetic::{
            conversions::{CONVERT_OPS_REGISTRY, EXTENSION},
            float_types::FLOAT64_TYPE,
            int_types::INT_TYPES,
        },
        type_row,
        types::{SumType, Type},
        Hugr,
    };
    use crate::emit::test::SimpleHugrConfig;
    use crate::custom::{float::add_float_extensions, int::add_int_extensions, prelude::add_default_prelude_extensions};

    fn test_conversion_op(name: impl AsRef<str>, in_type: Type, out_type: Type, int_width: u8) -> Hugr {
        SimpleHugrConfig::new()
            .with_ins(vec![in_type.clone()])
            .with_outs(vec![out_type.clone()])
            .with_extensions(CONVERT_OPS_REGISTRY.clone())
            .finish(|mut hugr_builder| {
                let [in1] = hugr_builder.input_wires_arr();
                let ext_op = EXTENSION
                    .instantiate_extension_op(
                        name.as_ref(),
                        [(int_width as u64).into()],
                        &CONVERT_OPS_REGISTRY
                    )
                    .unwrap();
                let outputs = hugr_builder
                    .add_dataflow_op(ext_op, [in1])
                    .unwrap()
                    .outputs();
                hugr_builder.finish_with_outputs(outputs).unwrap()
            })
    }

    #[rstest]
    fn test_convert(mut llvm_ctx: TestContext) -> () {
        let op_name = "convert_u";
        let width = 5;
        llvm_ctx.add_extensions(add_int_extensions);
        llvm_ctx.add_extensions(add_float_extensions);
        llvm_ctx.add_extensions(add_conversions_extension);
        let in_ty = INT_TYPES[width as usize].clone();
        let out_ty = FLOAT64_TYPE;
        let hugr = test_conversion_op(op_name.clone(), in_ty, out_ty, width);
        check_emission!(op_name, hugr, llvm_ctx);
    }

    #[rstest]
    #[case::trunc("trunc_u", 23.4, 5)]
    fn test_truncation(mut llvm_ctx: TestContext, #[case] op_name: &str, #[case] input: f64, #[case] width: u8) -> () {
        llvm_ctx.add_extensions(add_int_extensions);
        llvm_ctx.add_extensions(add_float_extensions);
        llvm_ctx.add_extensions(add_conversions_extension);
        llvm_ctx.add_extensions(add_default_prelude_extensions);
        let in_ty = FLOAT64_TYPE;
        let out_ty = SumType::new([
            INT_TYPES[width as usize].clone(),
            ERROR_TYPE,
        ]);
        let hugr = test_conversion_op(op_name.clone(), in_ty, out_ty.into(), width);
        check_emission!(op_name, hugr, llvm_ctx);
    }
}
