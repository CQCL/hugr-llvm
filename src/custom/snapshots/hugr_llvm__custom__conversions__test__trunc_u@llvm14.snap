---
source: src/custom/conversions.rs
expression: module.to_string()
---
; ModuleID = 'test_context'
source_filename = "test_context"

define { i32, { i32 }, { { i32, i8* } } } @_hl.main.1(double %0) {
alloca_block:
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  %1 = fptoui double %0 to i32
  %2 = insertvalue { i32 } undef, i32 %1, 0
  %3 = insertvalue { i32, { i32 }, { { i32, i8* } } } { i32 0, { i32 } poison, { { i32, i8* } } poison }, { i32 } %2, 1
  ret { i32, { i32 }, { { i32, i8* } } } %3
}
